src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+        // that didn't do size tracking.
+        basic_migration(
+            "CREATE TABLE registry_src (
+                registry_id INTEGER NOT NULL,
+                name TEXT NOT NULL,
+                size INTEGER,
+                timestamp INTEGER NOT NULL,
+                PRIMARY KEY (registry_id, name),
+                FOREIGN KEY (registry_id) REFERENCES registry_index (id) ON DELETE CASCADE
+             )",
+        ),
+        // Git db directories
+        basic_migration(
+            "CREATE TABLE git_db (
+                id INTEGER PRIMARY KEY AUTOINCREMENT,
+                name TEXT UNIQUE NOT NULL,
+                timestamp INTEGER NOT NULL
+             )",
+        ),
+        // Git checkout directories
+        basic_migration(
+            "CREATE TABLE git_checkout (
+                git_id INTEGER NOT NULL,
+                name TEXT UNIQUE NOT NULL,
+                size INTEGER,
+                timestamp INTEGER NOT NULL,
+                PRIMARY KEY (git_id, name),
+                FOREIGN KEY (git_id) REFERENCES git_db (id) ON DELETE CASCADE
+             )",
+        ),
+        // This is a general-purpose single-row table that can store arbitrary
+        // data. Feel free to add columns (with ALTER TABLE) if necessary.
+        basic_migration(
+            "CREATE TABLE global_data (
+                last_auto_gc INTEGER NOT NULL
+            )",
+        ),
+        // last_auto_gc tracks the last time auto-gc was run (so that it only
+        // runs roughly once a day for performance reasons). Prime it with the
+        // current time to establish a baseline.
+        Box::new(|conn| {
+            conn.execute(
+                "INSERT INTO global_data (last_auto_gc) VALUES (?1)",
+                [now()],
+            )?;
+            Ok(())
+        }),
+    ]
+}
+
+impl GlobalCacheTracker {
+    /// Creates a new [`GlobalCacheTracker`].
+    ///
+    /// The caller is responsible for locking the package cache with
+    /// [`CacheLockMode::DownloadExclusive`] before calling this.
+    pub fn new(config: &Config) -> CargoResult<GlobalCacheTracker> {
+        let mut conn = if config.cli_unstable().gc {
+            let db_path = Self::db_path(config);
+            // A package cache lock is required to ensure only one cargo is
+            // accessing at the same time. If there is concurrent access, we
+            // want to rely on cargo's own "Blocking" system (which can
+            // provide user feedback) rather than blocking inside sqlite
+            // (which by default has a short timeout).
+            let db_path =
+                config.assert_package_cache_locked(CacheLockMode::DownloadExclusive, &db_path);
+            Connection::open(db_path)?
+        } else {
+            // To simplify things (so there aren't checks everywhere for being
+            // enabled), just process everything in memory.
+            Connection::open_in_memory()?
```

Comment:
Should we still call `assert_package_cache_locked` in the "stable" code path so we're more likely to catch bugs when developing new features?  Or would we be concerned about that destablizing stable and just deal with it when we work to stabilize?